Static checks: adds a checker for dockerfiles

wainersm opened the pull request:
This will introduce a new checker to inspect the dockerfiles of the project with use of the [hadolint](https://github.com/hadolint/hadolint) tool.

Before turn on that check, it will be needed to delint all the dockerfiles. I already did it for the tests repository, remaining to change the files in the kata-containers repository.

Fixes #4149
Depends-on: github.com/kata-containers/kata-containers#3109

wainersm commented:
@jodh-intel @GabyCT, please, do not review the changes on the dockerfiles as I should test they are correct.

My intention opening this PR is to get early feedback from you.

I'm using the `hadolint` tool to check the files. It is the first time I use that tool but it seems very capable (e.g. it uses shell-checker to even inspect `RUN` scripts), well-maintained, and easy to install (statically built [releases](https://github.com/hadolint/hadolint/releases) or packaged in Fedora>=33). If you know of any other tool for that purpose, please let me know. I'd like to ear your opinion on two subjects in particular: the handler for templates (if we even should have such as handler) and the linter rules being ignored.

This is a work in progress as I still need to:
- somehow install the tool in the developer's workstation. I thought in pull the binary from https://github.com/hadolint/hadolint/releases and put into `~GOPATH/bin` (although I confess I think we overuse that directory).
- test I did not break any of the changed dockerfiles
- delint files on the kata-containers repository

jodh-intel commented on .ci/static-checks.sh:
```diff
@@ -1102,6 +1103,95 @@ static_check_json()
+			# It's known that the linter will fail to parse lines
+			# started with `@`. Also it might give false-positives
+		        # on some cases. Here we remove all markers as a best
+			# effort approache. If the template file is still
```
Nit: s/approache/approach/

jodh-intel reviewed (approved):
Thanks @wainersm.

A nit and a general suggestion to space out the code a bit more (1 blank line between blocks, functions, etc), but generally...

lgtm

wainersm commented:
@jodh-intel update the PR to include the handler for hadolint install.

Next:
- test my changes on dockerfiles doesn't break anything
- delint the kata-containers repository

jodh-intel commented on .ci/static-checks.sh:
```diff
@@ -1102,6 +1103,132 @@ static_check_json()
+	fi
+
+	if [ -z "$has_linter" ]; then
+		local download_url="${linter_url}/releases/download/v${linter_version}/hadolint-Linux-x86_64"
```
Please use `uname -m` rather than hard-coding architecture names.

wainersm replied:
btw, hadolint is only distributed to x86_64. Makes sense to skip the checker for non-x86 arches?

jodh-intel replied:
That's unfortunate, but skipping seems like the best approach at this stage. I presume there checking those files on a single architecture will still find all the lint issues anyhow too.

---
Aside: If you haven't seen the logo for `hadolint`, this might just make your day :smile::

- https://github.com/hadolint/hadolint#haskell-dockerfile-linter

Let's keep our dockerfiles litter free and fragrant y'all!

jodh-intel commented on Dockerfiles/BuildTest/Dockerfile:
```diff
@@ -3,7 +3,7 @@
 
 # Usage: FROM [image name]
-FROM busybox
+FROM busybox:1.34.1
```
I do wonder if we need a script that can check for updated docker images if we're going to hard-code versions. This version seems to include the latest CVE fixes in that package though.

wainersm replied:
As I mentioned in https://github.com/kata-containers/tests/pull/4194#issuecomment-979378569 we can ignore (in-line) the rules [DL3006](https://github.com/hadolint/hadolint/wiki/DL3006) and [DL3007](https://github.com/hadolint/hadolint/wiki/DL3007).

Actually `busybox` a good example of image that we don't expect breakage because of a version bumps, so being entitled to have those rules ignored. What do you think @jodh-intel ?

wainersm replied:
Maybe I should delete that dockerfile instead -> https://github.com/kata-containers/tests/issues/4222

jodh-intel commented on metrics/storage/blogbench_dockerfile/Dockerfile.in:
```diff
@@ -15,13 +15,17 @@ ENV BLOGBENCH_URL "https://download.pureftpd.org/pub/blogbench"
+	apt-get clean && \
+	rm -rf /var/lib/apt/lists/ && \
+	curl -OkL "${BLOGBENCH_URL}/blogbench-${BLOGBENCH_VERSION}.tar.gz" && \
+	tar xzf "blogbench-${BLOGBENCH_VERSION}.tar.gz -C /"
```
The quotes are a bit odd here - it would make more sense to me if the end quote finished at `.tar.gz"`.

wainersm replied:
True. I am going to fix it.

jodh-intel reviewed (commented):
Thanks for updating @wainersm. A few remaining comments.

GabyCT commented on kata-webhook/Dockerfile:
```diff
@@ -1,14 +1,14 @@
 #
 # SPDX-License-Identifier: Apache-2.0
-FROM golang:latest AS builder
+FROM golang:1.16-alpine3.14 AS builder
```
why we have a specific version here? can that be handle into the versions.yaml ?

wainersm replied:
The webhook app [builds on Golang 1.16](https://github.com/kata-containers/tests/blob/main/kata-webhook/go.mod).

I'm still a learner of Golang so I don't know the best way to handle it. We can stick with `latest` if you think it is the right thing. Wdyt @GabyCT ?

GabyCT commented on metrics/network/iperf3_dockerfile/Dockerfile:
```diff
@@ -3,13 +3,14 @@
 
 # Usage: FROM [image name]
-FROM fedora
+FROM fedora:34
```
why this version? why not latest?

wainersm replied:
If we use `latest` it will pick up Fedora 35 image. And I've seen dockerfile builds failing with Fedora 35 image just like below:

```
[root@xxxxx tests]# cd metrics/network/iperf3_dockerfile/
[root@xxxxx iperf3_dockerfile]# git diff
diff --git a/metrics/network/iperf3_dockerfile/Dockerfile b/metrics/network/iperf3_dockerfile/Dockerfile
index 2f6dce17..b7cc62f3 100644
--- a/metrics/network/iperf3_dockerfile/Dockerfile
+++ b/metrics/network/iperf3_dockerfile/Dockerfile
@@ -3,7 +3,7 @@
 # SPDX-License-Identifier: Apache-2.0
 
 # Usage: FROM [image name]
-FROM fedora:34
+FROM fedora:35
 
 # Version of the Dockerfile
 LABEL DOCKERFILE_VERSION="1.0"
[root@xxxxx iperf3_dockerfile]# podman build -t iperf3 .
STEP 1: FROM fedora:35
STEP 2: LABEL DOCKERFILE_VERSION="1.0"
--> abe43026566
STEP 3: RUN dnf -y update &&     dnf -y install iperf3 &&     dnf clean all
Fedora 35 - x86_64                              0.0  B/s |   0  B     00:00    
Errors during downloading metadata for repository 'fedora':
  - Curl error (6): Couldn't resolve host name for https://mirrors.fedoraproject.org/metalink?repo=fedora-35&arch=x86_64 [getaddrinfo() thread failed to start]
Error: Failed to download metadata for repo 'fedora': Cannot prepare internal mirrorlist: Curl error (6): Couldn't resolve host name for https://mirrors.fedoraproject.org/metalink?repo=fedora-35&arch=x86_64 [getaddrinfo() thread failed to start]
Error: error building at STEP "RUN dnf -y update &&     dnf -y install iperf3 &&     dnf clean all": error while running runtime: exit status 1

```

GabyCT commented on metrics/network/nuttcp_dockerfile/Dockerfile:
```diff
@@ -3,7 +3,7 @@
 
 # Usage: FROM [image name]
-FROM ubuntu
+FROM ubuntu:16.04
```
why 16.04? that is an old image from ubuntu

wainersm replied:
It seems huttcp image is never used by metrics (I grep'ed the logs of the metrics jobs and found nothing). So I was not sure how to test it with ubuntu 20.04. Because that files was created at the same time of [metrics/network/iperf_dockerfile/Dockerfile](https://github.com/kata-containers/tests/tree/main/metrics/network/iperf_dockerfile/Dockerfile) (which uses ubuntu 16.04 too) I thought 16.04 would be the safest value.

What version should I use instead? Or is it safe to keep using `latest`?

wainersm replied:
I changed it to use latest instead of 16.04.

GabyCT commented on metrics/storage/blogbench_dockerfile/Dockerfile.in:
```diff
@@ -5,7 +5,7 @@
 
 # Usage: FROM [image name]
-FROM @UBUNTU_REGISTRY@/ubuntu
+FROM @UBUNTU_REGISTRY@/ubuntu:20.04
```
why this particular version?:

wainersm replied:
Currently `ubuntu:latest` points to `ubuntu:20.04`. Should we stick with `latest`?

jodh-intel replied:
According to https://hub.docker.com/_/ubuntu?tab=description, `latest` == "latest LTS", which seems like a reasonable value here (shame the label isn't `latest-lts` though ;)

wainersm replied:
So for all Ubuntu images which I touched I can be using `latest` + ignore the rule  https://github.com/hadolint/hadolint/wiki/DL3007 ?

wainersm replied:
Here and in all Dockerfile under metrics I changed to use `latest`.

wainersm commented:
Updated this PR:
 - https://github.com/kata-containers/tests/pull/4194/commits/148528aeda7a2e2265f829b4e4d550b55b54447e contains the fix to quotation issue @jodh-intel pointed out.
 - added https://github.com/kata-containers/tests/pull/4194/commits/a8ad6ebfb5366b0d1443e6062b8d84379fecd801 to handle the files from the kata-containers repository
 - added  https://github.com/kata-containers/tests/pull/4194/commits/a46fb334f8ad52b101891655b8844535a6b6987a because hadolint is only released for x86_64 currently

GabyCT commented:
I am not a fan of having a hardcoded version in some of the Dockerfiles as it is hard to maintain

jodh-intel commented:
> I am not a fan of having a hardcoded version in some of the Dockerfiles as it is hard to maintain

@wainersm - I agree with @GabyCT. It always feels wrong/dirty to me I'm afraid. How about using that magic `envsubst(1)` so we can convert a `Dockerfile.in` into the final `Dockerfile` and pull the version from the `versions.yaml` database?

Frankly, we should do that for all the YAML files floating around that hardcode versions too.

wainersm commented:
> > I am not a fan of having a hardcoded version in some of the Dockerfiles as it is hard to maintain
> 
> @wainersm - I agree with @GabyCT. It always feels wrong/dirty to me I'm afraid. How about using that magic `envsubst(1)` so we can convert a `Dockerfile.in` into the final `Dockerfile` and pull the version from the `versions.yaml` database?
> 
> Frankly, we should do that for all the YAML files floating around that hardcode versions too.

First I'd like to check if I understood correctly your worries @GabyCT: are you not fan of pinning to a version? or that the pinned version should not be hard-coded in the Dockerfile? or both :)

The main reason why I suggested to use a linter for dockerfile is because using `latest` is error-prone. Recently [we had CI failing](https://github.com/kata-containers/kata-containers/issues/2960) because `fedora:latest` changed from 34 to 35. Not surprisingly hadolint has a [warning rule](https://github.com/hadolint/hadolint/wiki/DL3007) to catch use of `latest`. Nevertheless we should not follow that rule if it doesn't make sense for us; the tool allow us to disable rules completely or, add in-line comment on the line we want a rule ignored.

For example, if we find safe that `metrics/network/iperf3_dockerfile/Dockerfile` uses the latest fedora image then we can do:
```
diff --git a/metrics/network/iperf3_dockerfile/Dockerfile b/metrics/network/iperf3_dockerfile/Dockerfile
index 654471d4..1bed588e 100644
--- a/metrics/network/iperf3_dockerfile/Dockerfile
+++ b/metrics/network/iperf3_dockerfile/Dockerfile
@@ -3,6 +3,7 @@
 # SPDX-License-Identifier: Apache-2.0
 
 # Usage: FROM [image name]
+# hadolint ignore=DL3007
 FROM fedora
 
 # Version of the Dockerfile
```

jodh-intel commented:
@wainersm - We spent a lot of effort removing all hard-coded versions numbers and putting them into one of the two versions databases:

- https://github.com/kata-containers/kata-containers/blob/main/versions.yaml
- https://github.com/kata-containers/tests/blob/main/versions.yaml

This was required because we have a large body of scripts and configurations, and hard-coding versions became an unmangageable mess.

However, the well-structured YAML databases can be programatically queried and also programatically checked to see if there are upstream updates available. So we want to avoid hard-coding versions **anywhere** except those databases.

Dockerfiles and (ironically) yaml files have always been problematic so we've had to relax our views on those file types. But only because those file formats are "static". But if we used the `envsubst(1)` trick we can build those files from the `.in` versions of them, and get back to version sanity by storing the versions in the database files.

jodh-intel commented:
@wainersm - please can you rebase to resolve the conflict?

wainersm commented:
> @wainersm - please can you rebase to resolve the conflict?

@jodh-intel just rebased to resolve the conflict. Regarding the problem of leaving hard-coded versions on the dockerfiles, is it something you prefer that I tackle on this PR or on a follow up? My only concern about a follow up is to come up to the conclusion that hadolint won't be much of use because we will end up with some many templates that the tool can't parse.

jodh-intel commented:
@wainersm - let's leave the `envsubst(1)` changes for a follow-up PR.

wainersm commented:
Just updated this PR where I:
 * changed the Dockerfiles under `metrics` to use `ubuntu:latest` instead of a pined version

@GabyCT @jodh-intel could you please give another round of review? I think I addressed all your concerns and we agreed that hard-coded versions could be addressed on a follow up PR.

Notice that this PR still depends on https://github.com/kata-containers/kata-containers/pull/3109 that I should be fixing the issues.

GabyCT commented:
thanks @wainersm could you just please solve the conflicts?

wainersm commented:
> thanks @wainersm could you just please solve the conflicts?

Just rebased the code and resolved the conflicts. Thanks @GabyCT !

wainersm commented:
/test

jodh-intel commented:
/retest-arm
/retest-ubuntu-metrics
